From 5ba23214da820ae18fd4b9b7b2afd6f1f8be186d Mon Sep 17 00:00:00 2001
From: Shakked Zucker <shakkedz@campus.technion.ac.il>
Date: Mon, 25 May 2026 10:18:46 +0300
Subject: [PATCH] icmpv6: add Reflect All support for Extended Echo

Add support for the Reflect All ICMP extension object in ICMPv6
Extended Echo Requests.

When a supported Reflect All request is received, copy the request,
starting from its IPv6 header, into the object's payload. Then send an
ICMPv6 Extended Echo Reply with the object C-Type set to Reply - No
Error.

Add UAPI definitions for the Reflect All object class and its C-Type
values.

Signed-off-by: Shakked Zucker <shakkedz@campus.technion.ac.il>
---
 include/uapi/linux/icmpv6.h | 10 ++++
 net/ipv6/icmp.c             | 91 ++++++++++++++++++++++++++++++++++---
 2 files changed, 95 insertions(+), 6 deletions(-)

diff --git a/include/uapi/linux/icmpv6.h b/include/uapi/linux/icmpv6.h
index ecaece3..d2842cd 100644
--- a/include/uapi/linux/icmpv6.h
+++ b/include/uapi/linux/icmpv6.h
@@ -174,5 +174,15 @@ struct icmp6_filter {
 
 #define MLD2_ALL_MCR_INIT { { { 0xff,0x02,0,0,0,0,0,0,0,0,0,0,0,0,0,0x16 } } }
 
+struct icmp_ext_obj_hdr {
+	__be16	length;
+	__u8	class_num;
+	__u8	c_type;
+};
+
+#define ICMP_EXT_OBJ_CLASS_REFLECT_ALL			6
+#define ICMP_EXT_OBJ_CTYPE_REQUEST			0
+#define ICMP_EXT_OBJ_CTYPE_REPLY_NO_ERROR		1
+#define ICMP_EXT_OBJ_CTYPE_REPLY_UNSUPPORTED_OBJECT	2
 
 #endif /* _UAPI_LINUX_ICMPV6_H */
diff --git a/net/ipv6/icmp.c b/net/ipv6/icmp.c
index 1635da0..ced125a 100644
--- a/net/ipv6/icmp.c
+++ b/net/ipv6/icmp.c
@@ -69,6 +69,10 @@
 
 #include <linux/uaccess.h>
 
+#define ICMPV6_EXT_ECHO_REPLY_ACTIVE	0x04
+#define ICMPV6_EXT_ECHO_REPLY_IPV4	0x02
+#define ICMPV6_EXT_ECHO_REPLY_IPV6	0x01
+
 static DEFINE_PER_CPU(struct sock *, ipv6_icmp_sk);
 
 static int icmpv6_err(struct sk_buff *skb, struct inet6_skb_parm *opt,
@@ -722,10 +726,13 @@ static enum skb_drop_reason icmpv6_echo_reply(struct sk_buff *skb)
 	struct icmpv6_msg msg;
 	struct dst_entry *dst;
 	struct ipcm6_cookie ipc6;
+	struct sk_buff *reflect_skb = NULL;
 	u32 mark = IP6_REPLY_MARK(net, skb->mark);
 	SKB_DR(reason);
 	bool acast;
+	bool reflect_all = false;
 	u8 type;
+	unsigned int reply_len;
 
 	if (ipv6_addr_is_multicast(&ipv6_hdr(skb)->daddr) &&
 	    net->ipv6.sysctl.icmpv6_echo_ignore_multicast)
@@ -796,21 +803,93 @@ static enum skb_drop_reason icmpv6_echo_reply(struct sk_buff *skb)
 	ipc6.tclass = ipv6_get_dsfield(ipv6_hdr(skb));
 	ipc6.sockc.mark = mark;
 
-	if (icmph->icmp6_type == ICMPV6_EXT_ECHO_REQUEST)
-		if (!icmp_build_probe(skb, (struct icmphdr *)&tmp_hdr))
+	if (icmph->icmp6_type == ICMPV6_EXT_ECHO_REQUEST) {
+		size_t min_len = sizeof(struct icmp_ext_hdr) +
+				 sizeof(struct icmp_ext_obj_hdr);
+		struct icmp_ext_hdr *ext_hdr;
+		struct icmp_ext_obj_hdr *obj_hdr;
+		unsigned int obj_len;
+
+		if (!pskb_may_pull(skb, min_len))
+			goto out_dst_release;
+
+		icmph = icmp6_hdr(skb);
+		ext_hdr = (struct icmp_ext_hdr *)(icmph + 1);
+		obj_hdr = (struct icmp_ext_obj_hdr *)(ext_hdr + 1);
+		obj_len = ntohs(obj_hdr->length);
+
+		if (ext_hdr->version == 2 &&
+		    obj_hdr->class_num == ICMP_EXT_OBJ_CLASS_REFLECT_ALL) {
+			if (obj_hdr->c_type != ICMP_EXT_OBJ_CTYPE_REQUEST ||
+			    obj_len < sizeof(*obj_hdr))
+				goto out_dst_release;
+
+			if (!pskb_may_pull(skb, sizeof(*ext_hdr) + obj_len))
+				goto out_dst_release;
+
+			icmph = icmp6_hdr(skb);
+			reflect_all = true;
+		} else if (!icmp_build_probe(skb, (struct icmphdr *)&tmp_hdr)) {
+			goto out_dst_release;
+		}
+	}
+
+	if (reflect_all) {
+		struct icmp6hdr *reply_icmph;
+		struct icmp_ext_hdr *reply_ext_hdr;
+		struct icmp_ext_obj_hdr *reply_obj_hdr;
+		unsigned char *payload;
+		unsigned int obj_len;
+		unsigned int payload_len;
+		unsigned int copy_len;
+		unsigned int request_len;
+
+		reflect_skb = skb_copy(skb, GFP_ATOMIC);
+		if (!reflect_skb)
 			goto out_dst_release;
 
+		reply_icmph = icmp6_hdr(reflect_skb);
+		reply_ext_hdr = (struct icmp_ext_hdr *)(reply_icmph + 1);
+		reply_obj_hdr = (struct icmp_ext_obj_hdr *)(reply_ext_hdr + 1);
+		payload = (unsigned char *)(reply_obj_hdr + 1);
+
+		obj_len = ntohs(reply_obj_hdr->length);
+		payload_len = obj_len - sizeof(*reply_obj_hdr);
+		request_len = skb_tail_pointer(skb) - skb_network_header(skb);
+		copy_len = min_t(unsigned int, request_len, payload_len);
+
+		memcpy(payload, skb_network_header(skb), copy_len);
+
+		reply_obj_hdr->c_type = ICMP_EXT_OBJ_CTYPE_REPLY_NO_ERROR;
+		((u8 *)&tmp_hdr)[7] = ICMPV6_EXT_ECHO_REPLY_ACTIVE |
+					   ICMPV6_EXT_ECHO_REPLY_IPV6;
+
+		reply_ext_hdr->checksum = 0;
+		reply_ext_hdr->checksum = csum_fold(csum_partial(reply_ext_hdr,
+								 sizeof(*reply_ext_hdr) + obj_len,
+								 0));
+
+		reply_icmph->icmp6_type = ICMPV6_EXT_ECHO_REPLY;
+		reply_icmph->icmp6_cksum = 0;
+
+		msg.skb = reflect_skb;
+	}
+
+	reply_len = msg.skb->len + sizeof(struct icmp6hdr);
+
 	if (ip6_append_data(sk, icmpv6_getfrag, &msg,
-			    skb->len + sizeof(struct icmp6hdr),
-			    sizeof(struct icmp6hdr), &ipc6, &fl6,
+			    reply_len, sizeof(struct icmp6hdr), &ipc6, &fl6,
 			    (struct rt6_info *)dst, MSG_DONTWAIT)) {
 		__ICMP6_INC_STATS(net, idev, ICMP6_MIB_OUTERRORS);
 		ip6_flush_pending_frames(sk);
 	} else {
-		icmpv6_push_pending_frames(sk, &fl6, &tmp_hdr,
-					   skb->len + sizeof(struct icmp6hdr));
+		icmpv6_push_pending_frames(sk, &fl6, &tmp_hdr, reply_len);
 		reason = SKB_CONSUMED;
 	}
+
+	if (reflect_skb)
+		consume_skb(reflect_skb);
+
 out_dst_release:
 	dst_release(dst);
 out:
-- 
2.34.1

